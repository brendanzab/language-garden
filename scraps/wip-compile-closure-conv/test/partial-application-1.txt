let h : (Int -> Int) -> Int :=
  fun (x : Int -> Int) => x 7;

let g : Int :=
  fun (a : Int) =>
    let f : Int -> Int :=
      fun (b : Int) => a + b;
    h f;
--    ^ lambda lifting would fail here
--      too few arguments, so requires a closure
--      see: https://www.cse.chalmers.se/edu/year/2011/course/CompFun/lecture2.pdf#page=21

g
