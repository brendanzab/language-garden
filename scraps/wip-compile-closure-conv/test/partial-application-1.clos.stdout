let h : (Int -> Int) -> Int :=
  clos(fun (env : ()) (x : Int -> Int) => x 7, ());
let g : Int :=
  clos(
    fun (env : ((Int -> Int) -> Int,)) (a : Int) =>
      let f : Int -> Int :=
        clos(fun (env : (Int,)) (b : Int) => #add env.0 b, (a,));
      env.0 f,
    (h,)
  );
g
